File: 2196-reverse-nodes-in-even-length-groups/reverse-nodes-in-even-length-groups.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Function to reverse a linked list
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        prev = None
        while head:
            temp = head.next
            head.next = prev
            prev = head
            head = temp
        return prev
    
    def reverseEvenLengthGroups(self, head: ListNode) -> ListNode:
        # Creating a dummy node to avoid adding checks for the first node
        dummy = ListNode()
        dummy.next = head
        
        prev = dummy
        # Loop to determine the lengths of groups
        length = 1
        while head:
            tail = head
            next_head = None
            
            # Determining the length of the current group
            j = 1
            while j < length and tail and tail.next:
                tail = tail.next
                j += 1
            
            # Head of the next group
            next_head = tail.next
            
            if j % 2 == 0:
                # If even-sized group is found
                # Reversing the group and setting prev and head appropriately
                tail.next = None
                prev.next = self.reverseList(head)
                prev = head
                head.next = next_head
                head = next_head
            else:
                # If group is odd-sized, simply move to the next group
                prev = tail
                head = next_head
            
            # Increase the group size
            length += 1
        
        # Returning the head of the modified list
        return dummy.next

File: 2196-reverse-nodes-in-even-length-groups/reverse-nodes-in-even-length-groups.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* reverseList(ListNode* begin, ListNode* end) {
        ListNode* prev = begin, *cur = begin->next, *next=nullptr;
        while (cur != end) {
            next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }

        return prev;
    }

public:
    ListNode* reverseEvenLengthGroups(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode* begin = head->next, *tail = head, *end = begin, *tmp = nullptr;
        for (int k = 2; ; ++k) {
            int cnt = 0;
            for (int i = 0; i < k; ++i) {
                if (!end) {
                    break;
                }

                tmp = end;
                end = end->next;
                ++cnt;
            }

            if (cnt % 2 == 0) {
                tail->next = reverseList(begin, end);
                tail = begin;
            } else {
                tail->next = begin;
                tail = tmp;
            }

            begin = end;
            if (!begin) {
                tail->next = nullptr;
                break;
            }
        }

        return head;
    }
};